src/coreclr/jit/importer_vectorization.cpp

```diff
@@ -0,0 +1,741 @@
+    //                 [          value2          ]
+    //
+    UINT64   value2     = MAKEINT64(cns[len - 4], cns[len - 3], cns[len - 2], cns[len - 1]);
+    GenTree* firstIndir = impCreateCompareInd(this, data, TYP_LONG, dataOffset, value1);
+
+    ssize_t  offset      = dataOffset + len * sizeof(WCHAR) - sizeof(UINT64);
+    GenTree* secondIndir = impCreateCompareInd(this, gtClone(data), TYP_LONG, offset, value2);
+
+    // TODO-Unroll-CQ: Consider merging two indirs via XOR instead of QMARK
+    GenTreeColon* doubleIndirColon = gtNewColonNode(TYP_INT, secondIndir, gtNewFalse());
+    return gtNewQmarkNode(TYP_INT, firstIndir, doubleIndirColon);
+#else // TARGET_64BIT
+    return nullptr;
+#endif
+}
+
+//------------------------------------------------------------------------
+// impExpandHalfConstEquals: Attempts to unroll and vectorize
+//    Equals against a constant WCHAR data for Length in [8..32] range
+//    using either SWAR or SIMD. In a general case it will look like this:
+//
+//    bool equals = obj != null && obj.Length == len && (SWAR or SIMD)
+//
+// Arguments:
+//    data         - Pointer to a data to vectorize
+//    lengthFld    - Pointer to Length field
+//    checkForNull - Check data for null
+//    startsWith   - Is it StartsWith or Equals?
+//    cns          - Constant data (array of 2-byte chars)
+//    len          - Number of 2-byte chars in the cns
+//    dataOffset   - Offset for data
+//
+// Return Value:
+//    A pointer to the newly created SIMD node or nullptr if unrolling is not
+//    possible or not profitable
+//
+GenTree* Compiler::impExpandHalfConstEquals(
+    GenTree* data, GenTree* lengthFld, bool checkForNull, bool startsWith, WCHAR* cnsData, int len, int dataOffset)
+{
+    assert(len >= 0);
+
+    if (compCurBB->isRunRarely())
+    {
+        // Not profitable to expand
+        JITDUMP("impExpandHalfConstEquals: block is cold - not profitable to expand.\n");
+        return nullptr;
+    }
+
+    if (fgBBcount > 20)
+    {
+        // We don't want to unroll too much and in big methods
+        // TODO-Unroll-CQ: come up with some better heuristic/budget
+        JITDUMP("impExpandHalfConstEquals: method has too many BBs (>20) - not profitable to expand.\n");
+        return nullptr;
+    }
+
+    const genTreeOps cmpOp = startsWith ? GT_GE : GT_EQ;
+
+    GenTree* elementsCount = gtNewIconNode(len);
+    GenTree* lenCheckNode;
+    if (len == 0)
+    {
+        // For zero length we don't need to compare content, the following expression is enough:
+        //
+        //   varData != null && lengthFld == 0
+        //
+        lenCheckNode = gtNewOperNode(cmpOp, TYP_INT, lengthFld, elementsCount);
+    }
+    else
+    {
+        assert(cnsData != nullptr);
+
+        GenTree* indirCmp = nullptr;
+        if (len < 8) // SWAR impl supports len == 8 but we'd better give it to SIMD
+        {
+            indirCmp = impExpandHalfConstEqualsSWAR(gtClone(data), cnsData, len, dataOffset);
```

Comment:
Ditto on `data` and `gtClone`

Reply:
Changed data argument to be LclVar so it's cheap to clone by default now